dll/win32/kernel32/client/handle.c

```diff
@@ -187,6 +221,138 @@ DuplicateHandle(IN HANDLE hSourceProcessHandle,
+        // technically think we rely on a side effect of a failed call.. we get the name even though call failed
+        NtQueryObject(hSourceHandle, ObjectNameInformation, nameFull, sizeof(nameFull), &returnedLength);	
+    }
+    _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+    {
+        int excd = _SEH2_GetExceptionCode();
+        DPRINT1("DuplicateHandle: NtQueryObject generated an exception (%lx) for handle(%lx).\n", excd, hSourceHandle);			
+    }
+    _SEH2_END;
+    if ( wcscmp((LPWSTR)(nameFull + (2 * sizeof(WCHAR))), L"\\Device\\Afd") == 0 ) 
+    { // check if this belongs to AFD, if so it's a winsock socket
+        //DPRINT("DuplicateHandle: NtQueryObject returned (%ls) for handle(%lx).\n", (LPWSTR)(nameFull + (2 * sizeof(WCHAR))), hSourceHandle);			
+        DPRINT("DuplicateHandle: NtQueryObject winsock socket device(AFD) detected for handle(%lx).\n", hSourceHandle);				
+        // DynLink the winsock routines we need for this work.
+        DPRINT("DuplicateHandle: calling GetModuleHandleA(\"ws2_32\").\n");
+        winsock_module = GetModuleHandleA("ws2_32");
+        if (winsock_module)
+        {
+            DynWSASocket = (wsskcall)GetProcAddress(winsock_module, "WSASocketW");
+            getsockopt = (wsgsocall)GetProcAddress(winsock_module, "getsockopt");
+            if (getsockopt) 
+            {
+                // getsockopt is our gatekeeper. if it fails winsock access is not in a reliable state
+                int result = getsockopt((SOCKET)hSourceHandle, SOL_SOCKET, SO_TYPE, (char*)&optVal, &optLen);
+                if (result == 0) 
+                {
+                    if(optVal == SOCK_STREAM)
+                        DPRINT("(PID %lx) DuplicateHandle: socket is a TCP socket.\n");
+                    else if(optVal == SOCK_DGRAM)
+                        DPRINT("DuplicateHandle: socket is a UDP socket.\n");
+                    dupwinsock = (wsdskcall)GetProcAddress(winsock_module, "WSADuplicateSocketW");
+                    isSocket = TRUE; // we set this here which makes sockets fall back to normal handle duplication if we cannot dynlink
+                } else DPRINT1("Error calling getsockopt for inheritable socket.\n");
+            } else DPRINT1("Error GetProcAddressA() could not Dynlink getsockopt!\n");			
+        } else DPRINT1("Error GetModuleHandleA() could not find WS2_32 module handle.\n");
+        // NOTE: currently this code duplicates the handle/socket but DOES NOT CHAGE THE HANDLE ACCESS OR OPTIONS IN ANY WAY (ignores dwOption and access rights changes)
+        // TODO: revise code to respect access changes requested in parameters
+        if (isSocket && GetCurrentProcessId() == GetProcessId(hSourceProcessHandle)) 
+        {
+            canDupe = TRUE; // this is the only case that is fairly certain to work
+        }
+        if (isSocket && hGetParentProcessId() == GetProcessId(hSourceProcessHandle)) { // handle is from parent process, it can be directly duplicated if inheritable
```

Comment:
```suggestion
        else if (isSocket && hGetParentProcessId() == GetProcessId(hSourceProcessHandle)) { // handle is from parent process, it can be directly duplicated if inheritable
```